documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,192 @@
+ * value provided via the {@link EventStreamWriter#noteTime(long)} API for by any writer using that
+ * API at the current location in the stream.
+ *
+ * upperTimeBound will always be greater than or equal to lowerTimeBound.
+ */
+@Data
+public class TimeWindow {
+    private final long lowerTimeBound;
+    private final long upperTimeBound;
+}
+
+```
+
+Note that the `TimeWindow` that is seen by Readers can be affected by `timestampAggrigationTimeout` in the config above. Different applications may wish to be more or less conservative as to when Writers are 'alive' and need to be considered.
+
+As part of this watermarking feature many  internal API Changes (Server and Client) were made to support this new API.
+
+## Controller and Watermarking
+The Controller takes the `timestamps` from all of the Writers on the Stream and write out a `streamCut` that is greater than or equal to the maximum of all Writer’s positions and a `timestamp` that is the minimum of all of the Writers.
+
+A call to `noteTimestampFromWriter` or `commitTransaction` updates a structure with the position of the Writer (or where the Transaction committed) and the `timestamp` provided. In the case of commit the Controller obtains the position information itself, where as in the case of `noteTimestampFromWriter` it is explicitly provided by the client.
+
+To do this there needs to be a single Controller instance responsible for aggregating this data for any given Stream. It may be useful for this to be the same instance that is managing Transactions for that stream.
+
+In the event that a Controller instance crashes, the in memory structure will be temporarily lost. It can be recovered by reading the last value written. During recovery there will be a gap where time does not advance.
+
+There are two main objectives that was achieved:
+
+1. Recording Writer marks.
+2. Generating consolidated watermarks.
+
+Writers may contact any arbitrary Controller instance and report their `timestamps` and `Positions`. This Controller conditionally persists the reported value in the shared metadata store. For generating watermarks, all available Controller instances divide ownership of Streams amongst themselves such that each Controller instance is exclusively responsible for computing watermarks for only a subset of Streams. So for any Stream there is exactly for one Controller instance which is periodically generating marks by consolidating all the individual Writer marks which could be processed by any arbitrary Controller instance.
+
+### Writer Marks
+There are two types of Writers (Event Writers and Transaction Writers) each with their own characteristics with respect to time to position tracking. So Event Writers and Transaction Writers will have their corresponding marks generated differently. However, their marks will be reported at the same place in the shared metadata store described above.
```

Comment:
"However, their marks will be reported at the same place in the shared metadata store described above" - The important thing is how the timestamp shows up to the reader not where it is stored.